### PR TITLE
Feature/add reset state functionality

### LINKED ISSUE #136: We need a resetState that doesn't cause change events
We need to be able to give it a new state from a starting point.. for stopping and starting the app with a new state.. it doesn't mean the state _changed_ from what the app state was previously... it might just be a totally different app state.

### PR DESCRIPTION
**Pull request by TerrenceLJones**
This resolves #136.

### CODE REVIEW

File: test/unit/app-lifecycle.spec.js [hunk 1]
```diff
@@ -70,6 +70,25 @@ describe('App-Lifecycle', function() {
         expect(this.stopStub).to.have.not.been.calledTwice;
       });
     });
+
+    describe('and restarting the application', function() {
+      beforeEach(function() {
+        this.sinon.spy(this.myApp, 'initState');
+        this.myApp.restart();
```
Comment (paulfalgout) [inline test/unit/app-lifecycle.spec.js]: we should test passing state as an option to restart to make sure it changes

File: docs/app.md [hunk 2]
```diff
@@ -168,6 +169,12 @@ myApp.isRunning();
 myApp.start();
 ```
 
+### App `restart`
```
Comment (paulfalgout) [inline docs/app.md]: we need to document that this accepts `state` as an option as well as a code example

File: src/app.js [hunk 3]
```diff
@@ -158,24 +158,40 @@ const App = Marionette.Object.extend({
...
+   * @method restart
+   * @memberOf App
+   * @param {Object} [options] - Settings for the App passed through to events
+   * @returns {App}
+   */
+  restart(options) {
+    this.stop(options);
+    this.initState(options);
+    this.start(options);
+  },
```
Comment (TerrenceLJones) [inline src/app.js]: Also to remind myself...`restart` needs to return `this` for chaining purposes.

**Comment by TerrenceLJones** [comment]
@paulfalgout Please review changes.

File: src/mixins/state.js [hunk 4]
```diff
@@ -26,14 +26,14 @@ export default {
    */
   initState(options = {}) {
```
Comment (paulfalgout) [inline src/mixins/state.js]: I know this isn't exactly related to this PR, but `initState` should also `return this;` for consistency.  Not that _we_ want to, but then  `restart` could be `this.start(options).initState(options).stop(options);` which is something some people will enjoy.

**Pull request merged**